cmd/podman/images/scp.go

```diff
@@ -0,0 +1,344 @@
+		if err != nil {
+			return err
+		}
+		fmt.Println("Loaded images(s): " + strings.Join(rep.Names, ","))
+	}
+	return nil
+}
+
+// toRemote takes a dialed ssh client, and image information. it connects to the client
+// and copies the saved image dir over to the remote host and then loads it onto the machine
+// returns a string containing output or an error
+func savetoRemote(localFile string, tag string) (string, error) {
+	dial, remoteFile, err := createConnection(false)
+
+	if err != nil {
+		return "", err
+	}
+	defer dial.Close()
+
+	n, err := scpD.CopyTo(dial, localFile, remoteFile)
+	if err != nil {
+		errOut := (strconv.Itoa(int(n)) + " Bytes copied before error")
+		return " ", errors.Wrapf(err, errOut)
+	}
+	run := ""
+	if tag != "" {
+		return "", errors.Wrapf(define.ErrInvalidArg, "Renaming of an image is currently not supported")
+	}
+	podman := os.Args[0]
+	run = podman + " image load --input=" + remoteFile + ";rm " + remoteFile // run ssh image load of the file copied via scp
+	out, err := connection.ExecRemoteCommand(dial, run)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSuffix(out, "\n"), nil
+}
+
+// toRemote takes a dialed ssh client and image information. it connects to the client
+// and saves the specified image on the remote machine and then copies it to the specified local location
+// returns an error if one occurs.
+func savefromRemote(image, localFile string, tag string) error {
+	dial, remoteFile, err := createConnection(true)
+
+	if err != nil {
+		return err
+	}
+	defer dial.Close()
+
+	if tag != "" {
+		return errors.Wrapf(define.ErrInvalidArg, "Renaming of an image is currently not supported")
+	}
+	podman := os.Args[0]
+	run := podman + " image save " + image + " --format=oci-archive --output=" + remoteFile // run ssh image load of the file copied via scp. Files are reverse in thie case...
+	_, err = connection.ExecRemoteCommand(dial, run)
+	if err != nil {
+		return nil
+	}
+	n, err := scpD.CopyFrom(dial, remoteFile, localFile)
+	connection.ExecRemoteCommand(dial, "rm "+remoteFile)
+	if err != nil {
+		errOut := (strconv.Itoa(int(n)) + " Bytes copied before error")
+		return errors.Wrapf(err, errOut)
+	}
+	return nil
+}
+
+// makeRemoteFile creates the necessary remote file on the host to
+// save or load the image to. returns a string with the file name or an error
+func makeRemoteFile(dial *ssh.Client) (string, error) {
+	run := "mktemp"
+	remoteFile, err := connection.ExecRemoteCommand(dial, run)
+	if err != nil {
+		return "", err
+	}
+	remoteFile = strings.TrimSuffix(remoteFile, "\n")
+	if err != nil {
+		return "", err
+	}
+	return remoteFile, nil
+}
+
+// createConnections takes a boolean determining which ssh client to dial
+// and returns the dials client, its newly opened remote file, and an error if applicable.
+func createConnection(fromOrToRemote bool) (*ssh.Client, string, error) {
+	ind := 0
+	if !fromOrToRemote && len(scpOpts.Connections) > 1 {
```

Comment:
This is very indirect; a function responsible for creating a SSH connection shouldn’t need to make assumptions about the order in which save/load happens. Have the caller supply a single `URI` an `iden` value. (And that would make it easier to turn the URI/iden arrays into named source/destination parameters, eventually.)

`scp` now looks nice with no use of arrays and hard-coded indices, but only because they are tucked away in here.

---

The general direction I’m pointing towards (absolutely not a requirement for this review)
```go
type scpImage struct {
   remoteURI *URI // remote host that contains the image, or if not set, the image is local
   privateKeyPath string // valid only if remoteURI != NULL; a private key to use for connections to remoteURI, or ""
   imageName string // may be "" during parsing, but is never empty when returned from parseArgs
}

func parseImageArgument(arg string) (scpImage, error) { // can just return three values instead of an object
   res := scpImage{}
   if strings.Contains(arg, "::") {
       // parse the pre-:: part, turn it into (remoteURI, privateKeyPath), incl. the ServiceDestinations lookup and user mapping)
       imageName = // the after-:: part
   } else {
      imageName = arg
   }
   // validate imageName if not empty
}

func parseArgs(args []string) (scpImage, scpImage, error) {
   source = parseImageArgument(args[0])
   if len(args) > 1 {
     dest = parseImageArgument(args[1])
   else {
     dest = {remoteURI: nil, imageName: "" }
   }
   if source.imageName == "" { fail }
   if dest.imageName == "" { dest.imageName = source.imageName }
   return (source, dest)
}

func scp() {
   (source, dest) = parseArgs
  if source.ImageName != dest.ImageName { fail }
  tempFile := ioutil.TempFile()
  if source.RemoteURI != nil { // alternatively, a single helper function for savefromSource(source, tempFile)
    saveFromRemoteSource(source, tempFile)
  else {
    abiEngine.Save(source.ImageName, tempFile)
  }
  if dest.RemoteURI != nil { // alternatively, a single helper function…
    loadInRemoteDest(dest, tempFile)
  else {
    abiEngine.load(tempFile)
  }
}

func saveFromRemoteSource(source, tempFile) {
   dial := createConnection(source) // includes RemoteUI and ident
   …
}
```
, no 0/1 indices, clear separation between parsing/default values/rejected cases.

Again, I’m absolutely not asking for a restructuring like this, but at least keeping the 0/1 index knowledge restricted to the (image scp)-specific `parseArgs` and `scp`, and keeping the potentially reusable `createConnection` more general, would be nice.

Reply:
okay, will change this to be more general but then `saveToRemote` and `saveFromRemote` will need to call specific array values. Someone needs to use the values at some point since two connections can be specified.

Reply:
The above doesn’t have `connections[0], connections[1]`, it has `source, destination`; so we use names instead of indices, and let the compiler worry about memory allocation.

But, yeah, _given_ the existing `uri[], iden[]` arrays created by `parseArgs`, `scp` would call `saveFromRemote(uri[0], iden[0]`, like it used to do with `execCommand` a few versions back, and there would need to be two different `loadToRemote` calls with `[0]` or `[1]` indices, like there used to be. That’s not as clean as having named source/destination, but it’s localized to `parseArgs`+`scp`

Reply:
To be more explicit, the `[0]`/`[1]` indexing of `uri` should be in `scp`, not in `saveFromRemote`/`loadToRemote` either.

Reply:
I am interested in totally overhauling how the url and iden arrays are handled, but based on our convo last week with Dan, I think any overarching breaking changes should be for a v2 redesign which i'd be more than happy to spearhead.

For now, I changed the current structure to fit your criteria.